core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java

```diff
@@ -125,9 +135,74 @@ protected final void updateTopologyId(TopologyAffectedCommand command) throws In
       // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
       int newTopologyId = Math.max(currentTopologyId(), commandTopologyId + 1);
       command.setTopologyId(newTopologyId);
-      waitForTransactionData(newTopologyId);
-      return visitGetKeysInGroupCommand(context, command);
+      CompletableFuture<Void> transactionDataFuture = stateTransferLock.transactionDataFuture(newTopologyId);
+      return retryWhenDone(transactionDataFuture, newTopologyId, context, command, (ctx, cmd) -> visitGetKeysInGroupCommand(ctx, cmd));
+
+   }
+
+   protected <T extends VisitableCommand> CompletableFuture<Void> retryWhenDone(
+         CompletableFuture<Void> future, int topologyId, InvocationContext ctx, T command, RetryHandler<T> handler) throws Throwable {
+      if (future.isDone()) {
+         getLog().tracef("Retrying command %s for topology %d", command, topologyId);
+         return handler.handle(ctx, command);
+      } else {
+         CancellableRetry cancellableRetry = new CancellableRetry(command, topologyId, handler, ctx);
+         CompletableFuture<Void> retryFuture = future.thenComposeAsync(cancellableRetry, remoteExecutor);
+         // We want to time out the current command future, not the main topology-waiting future,
+         // but the command future can take longer time to finish.
+         ScheduledFuture<?> timeoutFuture = timeoutExecutor.schedule(() -> {
+            if (cancellableRetry.cancel()) {
+               retryFuture.completeExceptionally(new TimeoutException("Timed out waiting for topology " + topologyId));
+            }
+         }, transactionDataTimeout, TimeUnit.MILLISECONDS);
+         retryFuture.whenComplete((nil, throwable) -> timeoutFuture.cancel(false));
+         return retryFuture;
+      }
+   }
+
+   protected interface RetryHandler<T extends VisitableCommand> {
+      CompletableFuture<Void> handle(InvocationContext ctx, T command) throws Throwable;
    }
 
    protected abstract Log getLog();
+
+   private static class CancellableRetry<T extends VisitableCommand> implements Function<Void, CompletionStage<Void>> {
+      private static final AtomicIntegerFieldUpdater<CancellableRetry> cancellableRetryUpdater
+            = AtomicIntegerFieldUpdater.newUpdater(CancellableRetry.class, "cancelled");
+      private static final Log log = LogFactory.getLog(CancellableRetry.class);
+
+      private final T command;
+      private final int topologyId;
+      private final RetryHandler<T> handler;
+      private final InvocationContext ctx;
+      @SuppressWarnings("unused")
+      private volatile int cancelled = 0; // there's no AtomicBooleanFieldUpdater
```

Comment:
Hoe about using `AtomicInteger`? 0 - not canceled, 1 canceled. Just like boolean...